src/status_im2/common/keychain/events.cljs

```diff
@@ -86,26 +86,74 @@
   [key-uid password]
   (keychain/save-credentials key-uid key-uid (security/safe-unmask-data password) #()))
 
+(defn get-user-password!
+  [key-uid callback]
+  (keychain/get-credentials key-uid
+                            #(callback (when %
+                                         (-> %
+                                             (oops/oget "password")
+                                             (security/mask-data))))))
+
 (re-frame/reg-fx
  :keychain/get-user-password
  (fn [[key-uid callback]]
-   (keychain/get-credentials
-    key-uid
-    #(if % (callback (security/mask-data (oops/oget % "password"))) (callback nil)))))
+   (get-user-password! key-uid callback)))
 
 (rf/defn get-user-password
   [_ key-uid callback]
   {:keychain/get-user-password [key-uid callback]})
 
+(defn- password-migration-key-name
+  [key-uid]
+  (str key-uid "-password-migration"))
+
+(defn save-password-migration!
+  [key-uid]
+  (keychain/save-credentials
+   (password-migration-key-name key-uid)
+   key-uid
+   ;; NOTE: using the key-id as the password, but we don't really care about the
+   ;; value, we only care that it's there
+   key-uid
+   #(when-not %
+      (log/error
+       (str "Error while setting up keychain migration")))))
+
+(defn get-password-migration!
+  [key-uid callback]
+  (keychain/get-credentials
+   (password-migration-key-name key-uid)
+   (comp callback boolean)))
+
 (re-frame/reg-fx
  :keychain/clear-user-password
  (fn [key-uid]
+   (keychain/reset-credentials (password-migration-key-name key-uid))
    (keychain/reset-credentials key-uid)))
 
 (re-frame/reg-fx
  :keychain/save-password-and-auth-method
  (fn [{:keys [key-uid masked-password on-success on-error]}]
    (-> (save-user-password! key-uid masked-password)
        (.then #(save-auth-method! key-uid auth-method-biometric))
+       (.then #(save-password-migration! key-uid))
        (.then #(when on-success (on-success)))
        (.catch #(when on-error (on-error %))))))
+
+;; NOTE: migrating the plaintext password in the keychain
+;; with the hashed one. Added due to the sync onboarding
+;; flow, where the password arrives already hashed.
+(re-frame/reg-fx
+ :keychain/password-hash-migration
+ (fn [{:keys [key-uid callback]}]
+   (-> (get-password-migration! key-uid identity)
+       (.then (fn [migrated?]
+                (if migrated?
+                  (callback)
```

Comment:
If calling `callback` throws an exception, it will be swallowed and ignored since no other code is piping a `.catch` on the return value of the re-frame effect.

You can move the `.catch` outside the inner function so it also catches errors from calling `callback`, and then improving the error message if it has been migrated or not.

You may also use `try/catch` around `(callback)` if you want to assume that part of the code is synchronous. I prefer not to mix and assume that, but it should be fine because we don't use the return value of this re-frame effect.

Reply:
Thanks! Went for a default value for `callback` when destructuring instead so we don't do the check twice.